ethcore/light/src/on_demand/response_guard.rs

```diff
@@ -0,0 +1,167 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::time::Duration;
+use std::collections::HashMap;
+
+use failsafe;
+use super::{ResponseError, ValidityError};
+
+type ResponsePolicy = failsafe::failure_policy::SuccessRateOverTimeWindow<NoBackoff>;
+
+/// Response guard error type
+#[derive(Debug, Eq, PartialEq)]
+pub enum Error {
+	/// No majority, the error reason can't be determined
+	NoMajority,
+	/// Majority, with the error reason
+	Majority(Inner),
+}
+
+
+/// Dummy type to convert a generic type with no trait bounds
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, PartialOrd, Ord)]
+pub enum Inner {
+	/// Bad execution proof
+	BadProof,
+	/// RLP decoding
+	Decoder,
+	/// Empty response
+	EmptyResponse,
+	/// Wrong header sequence
+	HeaderByNumber,
+	/// Too few results
+	TooFewResults,
+	/// Too many results
+	TooManyResults,
+	/// Trie error
+	Trie,
+	/// Unresolved header
+	UnresolvedHeader,
+	/// No responses expected.
+	Unexpected,
+	/// Wrong hash
+	WrongHash,
+	/// Wrong Header sequence
+	WrongHeaderSequence,
+	/// Wrong response kind
+	WrongKind,
+	/// Wrong number
+	WrongNumber,
+	/// Wrong Trie Root
+	WrongTrieRoot,
+}
+
+/// Handle and register responses that can fail
```

Comment:
Maybe comments on the why we need a response handler would be interesting.

Ok I start to get an intuition of how it works. Basically we add this as a way to aggregate the values (response) for the circuit breaker.
Which would mean we want a global circuit breaker (which make sense considering the success strategy instead of error), and this is to handle results (so no backoff strategy needed).
My concern here is that the state machine use is not related with any other state machine (I may have not read the circuit breaker crate correctly), shouldn't the constructor should use a clone of an existing circuit breaker state?

Reply:
I have partly already answered this concern in another similar question.

This creates a separate `clean` state machine and should not be inferred by anything?
The entire object is moved once either `RequestGuard` `ResponseGuard` have reached `our failure state` or upon success!

Thoughts?

Reply:
You're right, both comments are really similar, so I reply in previous comment.